### ISSUE TITLE
Animation and rendering speed

### ISSUE THREAD

**Issue by zhong2001** [mks0601/ExAvatar_RELEASE#43]
Hello,

I am currently working on a project based on this repository, and I noticed that in Table A of your paper, it mentions that the driving and rendering rate can reach 26 fps on a V100. However, when I tested the following code on a GPU 3090, I only achieved 13 fps. I would like to understand whether the issue is related to the specific location or code I am using, or if it might be due to the GPU itself. I would appreciate your insights on this matter.

Thank you for your time and assistance.
```
        render_time = time.time()
        # forward
        with torch.no_grad():
            human_asset, human_asset_refined, human_offset, mesh_neutral_pose = tester.model.module.human_gaussian(
                smplx_param, cam_param)
            human_render = tester.model.module.gaussian_renderer(human_asset, render_shape, cam_param)
        renders_time = time.time() - render_time
        fps1 = 1 / renders_time
        print(f"Frame {frame_idx}: FPS: {fps1:.2f}")
```

**Comment by mks0601** [comment]
Hi, we can cache triplane features (https://github.com/mks0601/ExAvatar_RELEASE/blob/43e9fb9d5400c27c411c6e9b19efce95a19f3792/avatar/common/nets/module.py#L521) as they are constant features once the training is finished. I cached the triplane features when computing the running time.

**Comment by zhong2001** [comment]
Hi, I have been attempting to use the cached `tri_feat` in `animate.py` as shown below. 
```
        tri_feat = tester.model.module.human_gaussian.extract_tri_feature()
        render_time = time.time()
        # forward
        with torch.no_grad():
            human_asset, human_asset_refined, human_offset, mesh_neutral_pose = tester.model.module.human_gaussian(
                smplx_param, cam_param, tri_feat)
            human_render = tester.model.module.gaussian_renderer(human_asset, render_shape, cam_param)
        renders_time = time.time() - render_time
        fps1 = 1 / renders_time
        print(f"Frame {frame_idx}: FPS: {fps1:.2f}")
```
Accordingly, I modified the `forward` function in `module.py` to pass `tri_feat` as a parameter. 
```
    def forward(self, smplx_param, cam_param, tri_feat=None, is_world_coord=False):
        #other code
        if tri_feat is None:
            # extract triplane feature
            tri_feat = self.extract_tri_feature()
        else:
            tri_feat = tri_feat
            print('tri_feat_extracted')
```
As a result, the fps rate increased from 13 to 14, but it still doesn't reach 26. Could it be that I am using the caching method incorrectly?
![image](https://github.com/user-attachments/assets/a6817bee-efea-4755-808c-2e9e2ab2b049)

I would appreciate your guidance on this.

Thank you for your help.

**Comment by mks0601** [comment]
tri_feat is always not None as it has not been defined. use self.tri_feat

**Comment by zhong2001** [comment]
> tri_feat is always not None as it has not been defined. use self.tri_feat

I followed your suggestion and tried using `self.tri_feat` directly, but I encountered an error. Perhaps there is an issue with how I am using it. 
   ```
 def forward(self, smplx_param, cam_param, is_world_coord=False):
        # other code
        # extract triplane feature
        tri_feat = self.tri_feat
```
![image](https://github.com/user-attachments/assets/9cab8481-f289-4049-9349-ae9e31b3acee)

If possible, could you please provide a more specific code example? I would greatly appreciate your help.

**Comment by mks0601** [comment]
I mean you can initialize self.tri_feat=None and if tri_feat == None, you compute tri_feat and replace self.tri_feat with the computed one. For every forward, if self.tri_feat is not None, you can retrive the computed one.

**Comment by zhong2001** [comment]
Thank you for your reply, and I apologize for misunderstanding your suggestion.

You meant to add `self.tri_feat = None` in the class's `__init__(self) `method and then use the following code in `forward`, correct?
```
    def forward(self, smplx_param, cam_param, is_world_coord=False):
        mesh_neutral_pose, mesh_neutral_pose_wo_upsample, _, transform_mat_neutral_pose = self.get_neutral_pose_human(
            jaw_zero_pose=True, use_id_info=True)
        joint_zero_pose = self.get_zero_pose_human()

        if self.tri_feat is None:
            self.tri_feat = self.extract_tri_feature()
            tri_feat = self.tri_feat
            print('tri_feat_compute')
        else:
            tri_feat = self.tri_feat

        # get Gaussian assets
        geo_feat = self.geo_net(tri_feat)
        mean_offset = self.mean_offset_net(geo_feat)  # mean offset of Gaussians
        scale = self.scale_net(geo_feat)  # scale of Gaussians
        rgb = self.rgb_net(tri_feat)  # rgb of Gaussians
        mean_3d = mesh_neutral_pose + mean_offset  # 大 pose
```
If there are still any issues, I apologize for that, and I appreciate your continued help and patience.

Thank you once again.

**Comment by mks0601** [comment]
Yeap that's correct. I set the rendering resolution to 1024x1024 when measuring the speed. Also, you can turn off the pose conditioning for the faster speed as in the table of the paper.

**Comment by zhong2001** [comment]
Hello,

I have used the previous code, but there hasn't been a significant improvement in the overall speed.
![image](https://github.com/user-attachments/assets/a319bbe0-e68b-4c85-ba6c-c25b06c7467d)
 Additionally, I made the following modifications in the `main` code to ensure the rendering resolution is 1024x1024. 
```
#def main
    #render_shape = cv2.imread(osp.join(args.motion_path, 'frames', str(frame_idx_list[0]) + '.png')).shape[:2]
    render_shape = [1024,1024]
    
    for frame_idx in tqdm(frame_idx_list):
        #other code
        img = cv2.imread(osp.join(args.motion_path, 'frames', str(frame_idx) + '.png'))
        img = cv2.resize(img, (render_shape[1], render_shape[0]))
```
Unfortunately, there hasn't been any noticeable improvement in the speed, which is quite confusing to me.
![image](https://github.com/user-attachments/assets/9000f8ea-00b6-4bc3-9f68-76d1a6df9211)
Furthermore, I'm not entirely sure where exactly I should disable pose control. Would disabling it have any impact on the driving and rendering results?

If it's convenient for you, it would be great if you could share your test code. I would appreciate any guidance you can provide.

Thank you for your help.

**Comment by mks0601** [comment]
please post url of your forked repo

**Comment by zhong2001** [comment]
> 请发布您 fork 仓库的 URL

https://github.com/zhong2001/ExAvatar_RELEASE.git
I have integrated the code I modified into this link.

**Comment by mks0601** [comment]
Could you print a message at [here](https://github.com/zhong2001/ExAvatar_RELEASE/blob/80351d3d54014d5fd84164f4c3e872c324f31342/avatar/common/nets/module.py#L575) to ensure your implementation has no bugs? It seems okay for me. I guess the different running speed is from your gpu. Also, to turn off the pose conditioning, remove all 'refined'.

**Comment by zhong2001** [comment]
Hello,

Following your suggestion, I added print statements at [here](https://github.com/zhong2001/ExAvatar_RELEASE/blob/630ba3f8bb5c1a8c8c5bd9eac5627dce2e1b2a7a/avatar/common/nets/module.py#L575), and based on the results, it seems that there are no issues. It could perhaps be a GPU-related problem, so I plan to test it on another GPU in the near future.
![image](https://github.com/user-attachments/assets/ecd3c1b7-3fb1-47ac-81fb-fa1cbe85be32)
Additionally, I removed the "refined" part, but there doesn't seem to be any noticeable improvement in speed, which is another point that is puzzling me. For reference, the code where I removed "refined" can be found in [`animate_3.py` ](https://github.com/zhong2001/ExAvatar_RELEASE/blob/main/avatar/main/animate_3.py)and [`module2.py`](https://github.com/zhong2001/ExAvatar_RELEASE/blob/main/avatar/common/nets/module2.py).

I would appreciate any further insights you might have.

Thank you again for your help.

**Comment by mks0601** [comment]
simply not returning them does not make any changes. you can remove estimation parts to get refined to get speed boost.

**Comment by zhong2001** [comment]
You are right. Regarding the "refined" part, I have only modified the code in `module.py` so far. Should I also make changes in `avatar/main/model.py`? For example, something like this:https://github.com/mks0601/ExAvatar_RELEASE/blob/43e9fb9d5400c27c411c6e9b19efce95a19f3792/avatar/main/model.py#L186

**Comment by mks0601** [comment]
I mean you need to remove lines for estimating mean_offset_offset, scale_offset, and rgb_offset

**Comment by zhong2001** [comment]
Do you mean like this? https://github.com/zhong2001/ExAvatar_RELEASE/blob/8c158c7a742cfe6c28574815896136d3be285e8f/avatar/common/nets/module2.py#L595
However, the parts in [module2.py](https://github.com/zhong2001/ExAvatar_RELEASE/blob/main/avatar/common/nets/module2.py) that I provided earlier seem to already have these sections commented out, and there hasn't been a noticeable improvement in the results.

**Comment by mks0601** [comment]
remove
https://github.com/zhong2001/ExAvatar_RELEASE/blob/8c158c7a742cfe6c28574815896136d3be285e8f/avatar/common/nets/module2.py#L586
https://github.com/zhong2001/ExAvatar_RELEASE/blob/8c158c7a742cfe6c28574815896136d3be285e8f/avatar/common/nets/module2.py#L620

**Comment by zhong2001** [comment]
Thank you so much for your reply! It was indeed my oversight. I have now commented it out, and there has been a significant improvement in speed (14fps -> 26fps).

**Comment by AreinDaralnakhla** [comment]
can we view the final 3d model rather than just video? of the mesh along with the textures?

**Comment by mks0601** [comment]
we use 3D Gaussians, not the surface mesh representation.
